### PR TITLE
Allow game to specify first and last mod in mod loading order. 

### LINKED ISSUE #8934: Support for mod that is loaded before any other mod
## Problem

If you need to have code that must run before any other mod, there's currently not really any reliable method to do so.

Specifically, for overwriting things like `minetest.nodedef_defaults`, it is very crucial to have this ability. Only very early in the mod loading does it make sense to overwrite the node definition defaults. If you already had 2-3 mods before you had the ability to overwrite things, then it is probably already too late.

Not having this ability makes it practically impossible to overwrite `minetest.nodedef_defaults` (and similar) in a reasonable (!) way.

## Solution

Add support for a special mod (let's call it “`gameinit`”) that is treated specially by Minetest. If a mod with the name `gameinit` is found in the game, then Minetest will make sure this mod is always loaded before all other game mods and user-selected mods (but after `builtin`). Apart from that, this will be a mod like any other.

This mod *must* be a game mod, using the name for a regular mod must result in an error.

This will allow games to safely do their early initialization tasks like overwriting `minetest.nodedef_defaults`.

## Alternatives
Suggest them. :D

### PR DESCRIPTION
**Pull request by sfence**
Allow games to do something at the beginning and end of the mod loading process.

- Allows game creators to do something before all other mods are loaded and after all other mods are loaded.
- Use added parameter lines `first_mod` and `last_mod` in `game.conf` to determine which mod has to be loaded first and which has to be loaded last.
- This should fix #8934.
- This should help with the solution of issue #13442.

## To do

This PR is Ready for Review.

- [x] Get feedback
- [x] Apply feedback
- [x] Add documentation to lua_doc.md

## How to test

Switch to this branch and load the devtest game.
In devtest `game.conf` lines `first_mod` and `last_mod` are added. There are added checks in every devtest mod and in `last_mod` mod to verify if the loading order is right.

### CODE REVIEW

File: src/content/mod_configuration.cpp
Comment (appgurueu) [inline src/content/mod_configuration.cpp:241]: It looks like the formatting is a bit messy here

**Review by appgurueu — COMMENTED** [review]
(Unofficial review, not a core dev.)

Allowing games to do *something* here makes sense.

I'm not a fan of the repeated

```lua
if minetest.global_exists("first_mod") then
	table.insert(first_mod.mods,minetest.get_current_modname())
else
	error("ERROR: Mod "..minetest.get_current_modname().." loaded before first_mod.")
end
```

boilerplate in every devtest mod. Could you perhaps test this differently, for example by means of C++ unit tests?

**Comment by lhofhansl** [comment]
Alternatively we can add some hooks `before_mod_load`, `after_mod_load` (or something, not good at naming) that are executed at those stages.
Games can then implement these hooks in one or more of their mods. Also that way each mod can early and late things.

**Comment by sfence** [comment]
> Alternatively we can add some hooks `before_mod_load`, `after_mod_load` (or something, not good at naming) that are executed at those stages. Games can then implement these hooks in one or more of their mods. Also that way each mod can early and late things.

It will be complicated I think. Because every mod will have to be loaded twice time. At first time check if there are some hooks registration and the second time register nodes etc. So it will result in breaking backward compatibility, I am afraid.

But it will be a more flexible solution to do something like it.

**Comment by appgurueu** [comment]
> Also that way each mod can early and late things.

The problem with this is that it doesn't make much sense: There can't be multiple mods that load first or last. And this has practical implications: Mods that want to load first usually want to do so to, say, override some function. If you have two mods attempting to do this, you can easily get a conflict.

File: src/content/mod_configuration.cpp
Comment (sfence) [reply src/content/mod_configuration.cpp:241]: You ar right. Looks like spaces are used here as indent by mistake.

**Comment by sfence** [comment]
@appgurueu 

Thanks for the feedback.

Should be testable in C++ unittests. Lua solutions take less effort to create.
But C++ unittest will be probably better for it. I will look at it if I get in right mod for it.

**Comment by wsor4035** [comment]
just some a thought, not 100%

relating to the keys in game.conf, would it make sense to have them be something like `load_first_mod = modname` and `load_last_mod = modname` taking the syntax from world.mt

**Comment by sfence** [comment]
> just some a thought, not 100%
> 
> relating to the keys in game.conf, would it make sense to have them be something like `load_first_mod = modname` and `load_last_mod = modname` taking the syntax from world.mt

Not a problem to change key names.
But still waiting for roadmap approval or disapproval. So it does not make sense to continue work on this now.

**Comment by lhofhansl** [comment]
> The problem with this is that it doesn't make much sense: There can't be multiple mods that load first or last. And this has practical implications: Mods that want to load first usually want to do so to, say, override some function. If you have two mods attempting to do this, you can easily get a conflict.

Fair enough. I thoughts since games control their mods, they can control the sets of mods that do that.
If this is a better approach let's do this then.

**Comment by sfence** [comment]
> Fair enough. I thoughts since games control their mods, they can control the sets of mods that do that.
> If this is a better approach let's do this then.

I think, that `before` and `after` hooks can be implemented with this as well. It only requires the first loaded mod to go through mod lists and in every mod path looks for file `before_hook.lua` for example. `after` hook can be done also through callback registration. So it will be on the game developer to allow/disallow it.

**Comment by FreeLikeGNU** [comment]
Each world has a list of enabled mods.  Why not set the order with that list with enabled mods at the top of the list with the user able to change that order to needs? 

Slightly off-topic: It would be nice to have mod settings in that list as well instead of a just a boolean it could be something like:

##just use the mod's default or game.conf settings:
`load_mod_some_mod = true `
## this world uses this mod in a particular way:
`load_mod_some_mod = { settings = {good = true, bad = 11, ugly = "welcome"}`

**Comment by sfence** [comment]
> Each world has a list of enabled mods. Why not set the order with that list with enabled mods at the top of the list with the user able to change that order to needs?
> 
> Slightly off-topic: It would be nice to have mod settings in that list as well instead of a just a boolean it could be something like:
> 
> just use the mod's default or game.conf settings: `load_mod_some_mod = mods/some_mod `
> 
> this world uses this mod in a particular way: `load_mod_some_mod = { path= "mods/some_mod", settings = {good = true, bad = 11, ugly = "welcome" } }`

I am afraid, it will be extremely complicated for the user to determine the correct loading order to satisfy all dependencies in a manually configured loading order list. In addition, this feature works with an enabled mods list. It changes only the logic of ordering that list by dependencies.
What is the expected advantage of changing which mod will be loaded as 9th in order for example?

File: games/devtest/mods/last_mod/init.lua
Comment (rubenwardy) [inline games/devtest/mods/last_mod/init.lua:3]:
```suggestion
  loaded_mods = 0
```

File: src/content/mod_configuration.cpp
Comment (rubenwardy) [inline src/content/mod_configuration.cpp:242]:
```suggestion
			if (mod.name == last_mod) {
```

File: src/content/mod_configuration.cpp
Comment (rubenwardy) [inline src/content/mod_configuration.cpp:244]:
```suggestion
				satisfied.insert(it, mod);
```

File: src/content/mod_configuration.cpp
Comment (rubenwardy) [inline src/content/mod_configuration.cpp:246]:
```suggestion
				if (mod.name != first_mod) {
```

File: src/content/mod_configuration.cpp
Comment (rubenwardy) [inline src/content/mod_configuration.cpp:257]: Space used as indentation here, should be tabs. Make sure to set this in your editor. Also helps to have whitespace visible

**Review by rubenwardy — CHANGES_REQUESTED** [review]
You should update `src/unittest/test_servermodmanager.cpp` to add a test that first_mod and last_mod are in the correct places. If you do this, you could maybe remove all the checks from mods inside devtest

**Comment by rubenwardy** [comment]
Adding to 5.9.0 as it reduces the impact of #14486

File: src/content/mod_configuration.cpp
Comment (rubenwardy) [inline src/content/mod_configuration.cpp:256]: This should be checked as well at a different point. If you set the first_mod mod.conf to this, it will only break if mobs is  installed and enabled:

```
name = first_mod
description = Mod which should be loaded before every other mod.
optional_depends = mobs
```

File: src/content/mod_configuration.cpp
Comment (sfence) [reply src/content/mod_configuration.cpp]: Sorry. I am working on more projects, some of them using space, and some tabs indented. I change this setting often and sometimes I forget to do so. My fault.

File: src/content/mod_configuration.cpp
Comment (sfence) [reply src/content/mod_configuration.cpp]: Check moved.

File: src/content/mod_configuration.cpp
Comment (appgurueu) [inline src/content/mod_configuration.cpp:238]: Why have this restriction for the last mod? Since it loads last, it naturally can depend on anything (though that needn't be useful).

File: src/content/mod_configuration.cpp
Comment (appgurueu) [inline src/content/mod_configuration.cpp:199]: I think using empty strings for absent first / last mod is a bit dirty. Why not use `std::optional<std::string>` here to have some type safety?

File: src/content/mod_configuration.cpp
Comment (appgurueu) [inline src/content/mod_configuration.cpp:247]: Doing this in this loop is all a bit messy. It is not obvious why it works. Why not just:

* Prepend the "first" mod, if existent, to the mod load order
* Append the "last" mod, if existent, to the mod load order
* Ignore the "first" and "last" mod in this loop

(Also make sure to check that (1) nothing depends on the last mod and (2) the first mod depends on nothing.)

File: src/content/mod_configuration.cpp
Comment (appgurueu) [reply src/content/mod_configuration.cpp:199]: > shouldn't the `ModConfiguration` class store these actually? it seems to be responsible for sorting mods

Seems sensible. Might also simplify the code a bit by shortening parameter lists.
(Still the types should probably be `std::optional<std::string>`.)

File: src/content/mod_configuration.cpp
Comment (sfence) [reply src/content/mod_configuration.cpp]: Fixed.

File: src/content/mod_configuration.cpp
Comment (sfence) [reply src/content/mod_configuration.cpp]: Now first and last mod are astored in class. No additional parameters.

File: src/content/mod_configuration.cpp
Comment (sfence) [reply src/content/mod_configuration.cpp:266]: Fixed.

**Comment by sfence** [comment]
I have compiled Minetest on my MacBook A2 and run `./bin/minetest --run-unittests` command and I do not see the problem from the GitHub Macos test.
So hopefully, it is not a real problem.

Will be nice if somebody checks it also on a MacBook with an Intel CPU.

**Comment by grorp** [comment]
@fluxionary why 👎?

>  So hopefully, it is not a real problem.

I pressed the "run again" button, who knows. But if that doesn't work, you'll have to fix CI either way.

**Review by appgurueu — APPROVED** [review]
Thanks for your patience.

File: src/content/mod_configuration.h
Comment (sfan5) [inline src/content/mod_configuration.h:95]: fwiw I think it's okay to use an empty string as an implicit null, avoid possible confusion between `""` (often not valid) and `std::nullopt`

File: src/content/mod_configuration.cpp
Comment (sfan5) [inline src/content/mod_configuration.cpp:228]: btw the style guide says that variables are snake_case

File: src/content/mod_configuration.h
Comment (sfence) [reply src/content/mod_configuration.h:95]: Looks like @appgurueu should answer this.

Comment (appgurueu) [reply src/content/mod_configuration.h:95]: I could have left this as it was tbh, both are fine. Though since I already converted some to optionals, it improves consistency to also use optionals here. Strings are just used as mod "IDs" / "names" here. When I see a `std::string` in the context of a name / ID, I don't know if an empty string is allowed (a comment could help communicate this, but only by convention), and I can easily forget to handle the empty string properly.

`std::optional<std::string>` makes this pretty clear IMO if you know that the `std::string` is conceptually a name (and hence not supposed to be empty), though something like `std::optional<NonEmptyString>` would be better.

If we were to make this even cleaner, we could do `std::optional<ModName>` and `using ModName = std::string;`, and if we were to make this completely idiot-proof we could even have `ModName` be a type that enforces a valid, non-empty mod name (this would be overkill, though).

Comment (sfan5) [reply src/content/mod_configuration.h:95]: > `std::optional<std::string>` makes this pretty clear IMO if you know that the `std::string` is conceptually a name (and hence not supposed to be empty),

I would have interpreted that exactly opposite: someone went to the effort of adding `std:optional<>`, so there must be something special about `""` that makes it different from a null value. maybe it's valid too?

Comment (sfence) [reply src/content/mod_configuration.h:95]: So, should I revert the last commit (std::optional add)?

`fistModSpec` and `lastModSpec` naming is now fixed in the fist commit also.

Comment (sfan5) [reply src/content/mod_configuration.h:95]: @appgurueu opinion?

Comment (appgurueu) [reply src/content/mod_configuration.h:95]: Seeing as it seems to be considered counterproductive, let me undo the optional-ification of these two strings.

File: src/content/mod_configuration.cpp
Comment (sfan5) [inline src/content/mod_configuration.cpp:269]:
```suggestion
			throw ModError("Impossible to depend on the mod specified by last_mod");
```

File: src/content/mod_configuration.cpp
Comment (sfan5) [inline src/content/mod_configuration.cpp:235]: does this actually do anything? I see this being overriden in the loop below

perhaps `mod.optdepends.clear();` would make more sense

File: src/content/mod_configuration.cpp
Comment (sfan5) [inline src/content/mod_configuration.cpp:239]: what's the significance of removing first and last mod from `modnames`?

AFAICT this means nobody can depend on them, which makes sense for last_mod. but not necessarily for first_mod

File: src/content/mod_configuration.cpp
Comment (sfan5) [inline src/content/mod_configuration.cpp:283]:
```suggestion
			throw ModError("Mod specified by first_mod cannot have dependencies");
```

File: src/content/mod_configuration.cpp
Comment (sfan5) [inline src/content/mod_configuration.cpp:314]: can't this just be written to `m_unsatisfied_mods` and left to normal error handling?

File: src/content/mod_configuration.h
Comment (sfan5) [inline src/content/mod_configuration.h:23]:
```suggestion
```

File: src/content/subgames.h
Comment (sfan5) [inline src/content/subgames.h:26]:
```suggestion
```

**Review by sfan5 — CHANGES_REQUESTED** [review]

File: src/content/mod_configuration.cpp
Comment (appgurueu) [reply src/content/mod_configuration.cpp:314]: that should work

File: src/content/mod_configuration.cpp
Comment (appgurueu) [reply src/content/mod_configuration.cpp:235]: this seems to be unnecessary indeed; i think clearing the optional depends shouldn't be necessary either: if the opt depends are there, then they will be added to the depends, and hence processed before the last mod and satisfied.

i think the correct thing to do is to just remove the line

File: src/content/mod_configuration.cpp
Comment (appgurueu) [reply src/content/mod_configuration.cpp:237]: This is used just for optional depends, so it's fine to not add it for the first mod - if any mods optionally depend on it, these optional depends are automatically fulfilled.

**Comment by appgurueu** [comment]
While working on this a bit, I realized that we apparently support-in world games with some duplicated code, so I de-duplicated that and added support for first/last mod there too: https://github.com/minetest/minetest/pull/14177/commits/a7d79dc80bed7b54e5a8bda4a77515a00e6c46ae. I tested this briefly by creating a new world by copying an existing world, editing `world.mt` to the new name and copying devtest to the `game` subfolder of the new world - it seems to work as expected.

**Comment by sfan5** [comment]
> ps: it seems the `TestServerModManager` is skipped in CI tests

where do see that? it's literally in the log you linked.

**Review by sfan5 — APPROVED** [review]

**Comment by appgurueu** [comment]
> > ps: it seems the `TestServerModManager` is skipped in CI tests
> 
> where do you see that? it's literally in the log you linked.

Okay, apparently my browser is broken:

Screenshot from 2024-06-01 00-50-54

(The reason I initially believed CI to be skipping tests was that CI passed, while my tests locally failed. I then "confirmed" this by Ctrl+F'ing `TestServerModManager`. The real reason for the discrepancy was that I had accidentally used an older version of devtest because I was not using `RUN_IN_PLACE=1`.)

**Comment by jordan4ibanez** [comment]
If you don't merge this I'm gonna have to post a gif of a cat crying

**Pull request merged**
